File: exercises/F/exercise3/template/fractal-set.f90
program mandelbrot
  use mpi
  use write_image
  use read_opt

  implicit none

  abstract interface
     integer function fn(c, max_iter)
       complex, intent(in) :: c
       integer, intent(in) :: max_iter
     end function fn
  end interface

  integer, allocatable, dimension(:,:) :: image

  integer :: gridsizex, gridsizey
  integer :: iter
  integer :: ierr
  integer :: rank
  integer :: comm
  complex :: cmin
  complex :: cmax

  procedure(fn), pointer :: fp

  call MPI_Init(ierr)
  comm = MPI_COMM_WORLD
  call MPI_Comm_rank(comm, rank, ierr)
  call read_options(gridsizex, gridsizey, iter, cmin, cmax)

  if ( rank .eq. 0 ) then
     call initialise_image(image, gridsizex, gridsizey)
  end if

  ! Set fp to point to the correct function (point_in_mandelbrot_set
  ! or point_in_julia_set)

  call compute_set(fp, image, cmin, cmax, gridsizex, gridsizey, iter)

  if ( rank .eq. 0 ) then
     call write_ppm("output.ppm", image, iter)
     deallocate(image)
  end if

  call MPI_Finalize(ierr)

contains

  ! Initialise the data for the image array.  Data is stored in
  ! "scanline order", i.e. x dimension varies fastest.  You get an
  ! array with shape (grid_size_x, grid_size_y) from this function
  subroutine initialise_image(image, grid_size_x, grid_size_y)
    integer, intent(out), allocatable, dimension(:,:) :: image
    integer, intent(in) :: grid_size_x, grid_size_y

    allocate(image(grid_size_x, grid_size_y))
    image = -1
  end subroutine initialise_image

  subroutine calc_slice_bounds(slice, nslice, grid_size_y, &
       start, end)
    integer, intent(in) :: slice, nslice, grid_size_y
    integer, intent(out) :: start, end

    start = 1 + ((slice - 1) * grid_size_y / nslice)
    end = 1 + (slice * grid_size_y / nslice)
  end subroutine calc_slice_bounds

  subroutine send_image_slice(image_slice, start, end, dest, comm)
    integer, intent(in), dimension(:,:) :: image_slice
    integer, intent(in) :: dest
    integer, intent(in) :: comm
    integer, intent(in) :: start, end

    integer :: ierr

    call MPI_Send(start, 1, MPI_INTEGER, dest, 0, comm, ierr)
    call MPI_Send(end, 1, MPI_INTEGER, dest, 0, comm, ierr)
    call MPI_Send(image_slice, size(image_slice, 1) * (end - start), &
         MPI_INTEGER, dest, 0, comm, ierr)
  end subroutine send_image_slice

  subroutine recv_image_slice(image, source, comm)
    integer, intent(inout), dimension(:,:) :: image
    integer, intent(in) :: source
    integer, intent(in) :: comm
    
    integer :: start, end, ierr
    integer, dimension(MPI_STATUS_SIZE) :: status

    call MPI_Recv(start, 1, MPI_INTEGER, source, 0, comm, status, ierr)
    call MPI_Recv(end, 1, MPI_INTEGER, source, 0, comm, status, ierr)
    call MPI_Recv(image(1, start), size(image, 1) * (end - start), &
         MPI_INTEGER, source, 0, comm, status, ierr)

  end subroutine recv_image_slice

  subroutine copy_slice_to_image(image_slice, image, grid_size_y, &
       slice, nslice)
    integer, dimension(:,:), intent(in) :: image_slice
    integer, dimension(:,:), intent(inout) :: image
    integer, intent(in) :: slice, nslice, grid_size_y

    integer :: comm, rank, nproc, ierr
    integer :: start, end, i

    comm = MPI_COMM_WORLD
    call MPI_Comm_rank(comm, rank, ierr)

    if ( rank .eq. 0 ) then
       ! Copy local slices into global image
       ! Receive remote slices into global image
    else
       ! Send slices to rank 0
    end if
  end subroutine copy_slice_to_image

  pure function point_in_mandelbrot_set(c, max_iter) result(ret)
    complex, intent(in) :: c
    integer, intent(in) :: max_iter
    integer :: ret
    ret = 0
  end function point_in_mandelbrot_set

  pure function point_in_julia_set(c, max_iter) result(ret)
    complex, intent(in) :: c
    integer, intent(in) :: max_iter
    integer :: ret
    ret = 0
  end function point_in_julia_set

  subroutine compute_slice(in_set_fn, image_slice, slice, nslice, &
       min, max, grid_size_x, grid_size_y, max_iter)
    procedure(fn) :: in_set_fn
    integer, allocatable, dimension(:,:), intent(out) :: image_slice
    integer, intent(in) :: slice, nslice
    complex, intent(in) :: min, max
    integer, intent(in) :: grid_size_x, grid_size_y
    integer, intent(in) :: max_iter

    ! Compute the set given by IN_SET_FN in a SLICE of the whole domain.
    ! The total number of slices is given by NSLICE.
    !
    ! The utility function calc_slice_bounds is provided to work out
    ! the bounds of a slice.  It divides the image in the y direction.
    !
    ! This subroutine should return a newly initialised image slice.
  end subroutine compute_slice

  subroutine compute_set(in_set_fn, image, min, max, grid_size_x, &
       grid_size_y, max_iter)
    integer, dimension(:,:), intent(inout) :: image
    procedure(fn) :: in_set_fn
    complex, intent(in) :: min, max
    integer, intent(in) :: grid_size_x, grid_size_y
    integer, intent(in) :: max_iter
    integer :: rank, nproc, comm, ierr
    integer :: slice
    integer :: nslice
    integer, allocatable, dimension(:,:) :: image_slice

    comm = MPI_COMM_WORLD
    call MPI_Comm_rank(comm, rank, ierr)
    call MPI_Comm_size(comm, nproc, ierr)

    nslice = nproc
    slice = rank + 1
    call compute_slice(in_set_fn, image_slice, slice, nslice, &
         min, max, grid_size_x, grid_size_y, max_iter)

    call copy_slice_to_image(image_slice, image, &
         grid_size_y, slice, nslice)
    if ( allocated(image_slice) ) deallocate(image_slice)
  end subroutine compute_set

end program mandelbrot

File: exercises/C/exercise3/solution/fractal-set.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <error.h>
#include <mpi.h>
#include "arralloc.h"
#include "write_ppm.h"
#include "read_options.h"

typedef int (*in_set_fn_t)(const float, const float, const int);

static inline int point_in_mandelbrot_set(const float x0,
                                          const float y0,
                                          const int max_iter)
{
    int i;
    float x;
    float y;
    float x2;
    float y2;
    x = x0;
    y = y0;
    for ( i = 0; i < max_iter; i++ ) {
        x2 = x * x;
        y2 = y * y;
        /* z = (z*z) + c */
        if ( x2 + y2 > 4.0 ) {
            return i;
        } else {
            y = y0 + (2.0 * x * y);
            x = x0 + x2 - y2;
        }
    }
    return max_iter;
}

static inline int point_in_julia_set(const float x0,
                                     const float y0,
                                     const int max_iter)
{
    int i;
    float x;
    float y;
    const float cim = 0.01;
    const float cre = 0.285;
    x = x0;
    y = y0;
    for ( i = 0; i < max_iter; i++ ) {
        if ( x * x + y * y > 4.0 ) {
            return i;
        } else {
            float tmp = y * y;
            y = 2 * x * y + cim;
            x = x * x - tmp + cre;
        }
    }
    return max_iter;
}

/* Initialise data for image array.  Data is stored in "scanline
 * order", i.e. x dimension varies fastest.  You get an array with
 * shape image[grid_size_y][grid_size_x] from this function. */
void initialise_image(int ***image, const int grid_size_x, const int grid_size_y)
{
    int i;
    int j;
    *image = (int**)arralloc(sizeof(int), 2, grid_size_y, grid_size_x);

    if ( NULL == *image ) {
        error(1, errno, "Unable to allocate memory for image\n");
    }
    /* initalise results array to black */
    for ( i = 0; i < grid_size_y; i++ ) {
        for ( j = 0; j < grid_size_x; j++ ) {
            (*image)[i][j] = -1;
        }
    }
}

static inline void calc_slice_bounds(const int slice, const int nslice,
                                     const int grid_size_y,
                                     int *start, int *end)
{
    *start = slice * grid_size_y / nslice;
    *end = (slice + 1) * grid_size_y / nslice;
}

static inline void send_image_slice(int **image_slice,
                                    int start,
                                    int end,
                                    const int grid_size_x,
                                    int dest,
                                    MPI_Comm comm)
{
    MPI_Send(&start, 1, MPI_INT, dest, 0, comm);
    MPI_Send(&end, 1, MPI_INT, dest, 0, comm);
    MPI_Send(&(image_slice[0][0]), grid_size_x * (end - start), MPI_INT,
             dest, 0, comm);
}

static inline void recv_image_slice(int **image,
                                    const int grid_size_x,
                                    int source,
                                    MPI_Comm comm)
{
    int start;
    int end;
    MPI_Status status;
    MPI_Recv(&start, 1, MPI_INT, source, 0, comm, &status);
    MPI_Recv(&end, 1, MPI_INT, source, 0, comm, &status);
    MPI_Recv(&(image[start][0]), grid_size_x * (end - start), MPI_INT,
             source, 0, comm, &status);
}

void copy_slice_to_image(int **image_slice, int **image,
                         const int slice, const int nslice,
                         const int grid_size_x, const int grid_size_y)
{
    int start;
    int end;
    int rank;
    MPI_Comm comm;

    comm = MPI_COMM_WORLD;
    MPI_Comm_rank(comm, &rank);

    /* We send these values to master process for writing because we
     * don't assume that the mapping from rank to slice is
     * one-to-one.  For example, we might divide the domain into
     * twice as many slices as there are processes. */
    calc_slice_bounds(slice, nslice, grid_size_y, &start, &end);

    if ( rank == 0 ) {
        int i;
        int j;
        int size;

        /* Copy local bit into global image */
        for ( i = start; i < end; i++ ) {
            for ( j = 0; j < grid_size_x; j++ ) {
                image[i][j] = image_slice[i - start][j];
            }
        }

        /* Receive remote slices */
        MPI_Comm_size(comm, &size);
        for ( i = 1; i < size; i++ ) {
            recv_image_slice(image, grid_size_x, i, comm);
        }
    } else {
        send_image_slice(image_slice, start, end, grid_size_x, 0, comm);
    }
}

int **compute_slice(in_set_fn_t in_set_fn,
                    const int slice, const int nslice,
                    const float xmin, const float xmax,
                    const float ymin, const float ymax,
                    const int grid_size_x, const int grid_size_y,
                    const int max_iter)
{
    int i;
    int j;
    int start;
    int end;
    int **image_slice;
    float x0;
    float y0;

    calc_slice_bounds(slice, nslice, grid_size_y, &start, &end);

    initialise_image(&image_slice, grid_size_x, end - start);

    for ( i = 0; i < end - start; i++ ) {
        for ( j = 0; j < grid_size_x; j++ ) {
            /* calculate coordinates x0,y0 of current pixel */
            x0 = xmin + j * ((xmax - xmin) / grid_size_x);
            y0 = ymin + (i + start) * ((ymax - ymin) / grid_size_y);
            image_slice[i][j] = (*in_set_fn)(x0, y0, max_iter);
        }
    }
    return image_slice;
}

void compute_set(in_set_fn_t in_set_fn,
                 int **image,
                 const float xmin,
                 const float xmax,
                 const float ymin,
                 const float ymax,
                 const int grid_size_x,
                 const int grid_size_y,
                 const int max_iter)
{
    int rank;
    int size;
    int slice;
    int nslice;
    int **image_slice;
    MPI_Comm comm;

    comm = MPI_COMM_WORLD;
    MPI_Comm_rank(comm, &rank);
    MPI_Comm_size(comm, &size);

    nslice = size;
    slice = rank;
    /* Compute local image slice */
    image_slice = compute_slice(in_set_fn, slice, nslice, xmin, xmax, ymin, ymax,
                                grid_size_x, grid_size_y, max_iter);

    /* Copy into global image, note IMAGE is only defined on rank 0 */
    copy_slice_to_image(image_slice, image, slice, nslice, grid_size_x, grid_size_y);

    free(image_slice);
}

int main(int argc, char** argv)
{
    int grid_size_x;
    int grid_size_y;
    int max_iter;
    float xmin;
    float xmax;
    float ymin;
    float ymax;
    int **image;
    int rank;
    MPI_Comm comm;
    in_set_fn_t fp;

    MPI_Init(&argc, &argv);

    comm = MPI_COMM_WORLD;
    MPI_Comm_rank(comm, &rank);

    read_options(argc, argv, &grid_size_x, &grid_size_y, &max_iter,
                 &xmin, &xmax, &ymin, &ymax);

    if ( rank == 0 ) {
        initialise_image(&image, grid_size_x, grid_size_y);
    }

    fp = &point_in_julia_set;

    compute_set(fp, image,
                xmin, xmax, ymin, ymax,
                grid_size_x, grid_size_y, max_iter);

    if ( rank == 0 ) {
        write_ppm("output.ppm", image, grid_size_x, grid_size_y, max_iter);
        free(image);
    }

    MPI_Finalize();
    return 0;
}
